eZ/Publish/Core/Repository/URLService.php

```diff
@@ -0,0 +1,232 @@
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function updateUrl(URL $url, URLUpdateStruct $struct)
+    {
+        if ($this->repository->hasAccess('url', 'update') !== true) {
+            throw new UnauthorizedException('url', 'update');
+        }
+
+        if (!$this->isUnique($url->id, $struct->url)) {
+            throw new InvalidArgumentException('struct', 'url already exists');
+        }
+
+        $updateStruct = $this->buildUpdateStruct($this->loadById($url->id), $struct);
+
+        $this->repository->beginTransaction();
+        try {
+            $this->urlHandler->updateUrl($url->id, $updateStruct);
+            $this->repository->commit();
+        } catch (\Exception $e) {
+            $this->repository->rollback();
+            throw $e;
+        }
+
+        return $this->loadById($url->id);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadById($id)
+    {
+        if ($this->repository->hasAccess('url', 'view') !== true) {
+            throw new UnauthorizedException('url', 'view');
+        }
+
+        return $this->buildDomainObject(
+            $this->urlHandler->loadById($id)
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadByUrl($url)
+    {
+        if ($this->repository->hasAccess('url', 'view') !== true) {
+            throw new UnauthorizedException('url', 'view');
+        }
+
+        return $this->buildDomainObject(
+            $this->urlHandler->loadByUrl($url)
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createUpdateStruct()
+    {
+        return new URLUpdateStruct();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findUsages(URL $url, $offset = 0, $limit = -1)
+    {
+        $usages = $this->urlHandler->findUsages($url->id);
+
+        $query = new Query();
+        if (!empty($usages)) {
+            $query->filter = new ContentCriterion\LogicalAnd([
+                new ContentCriterion\ContentId($usages),
+                new ContentCriterion\Visibility(ContentCriterion\Visibility::VISIBLE),
+            ]);
+        } else {
+            $query->filter = new ContentCriterion\MatchNone();
+        }
+
+        $query->offset = $offset;
+        if ($limit > -1) {
+            $query->limit = $limit;
+        }
+
+        return $this->repository->getSearchService()->findContentInfo($query);
+    }
+
+    protected function buildDomainObject(SPIUrl $data)
+    {
+        return new URL([
+            'id' => $data->id,
+            'url' => $data->url,
+            'originalUrlMd5' => $data->originalUrlMd5,
+            'isValid' => $data->isValid,
+            'lastChecked' => $this->createDateTime($data->lastChecked),
+            'created' => $this->createDateTime($data->created),
+            'modified' => $this->createDateTime($data->modified),
+        ]);
+    }
+
+    protected function buildUpdateStruct(URL $url, URLUpdateStruct $data)
```

Comment:
Phpdoc please.

Reply:
Done in 0faac18c2e88bcf94b44f3452104d681c3ea9ecb.